### ISSUE TITLE
MPI and OpenMP threads

### ISSUE THREAD

**Issue by afrawake** [SmileiPIC/Smilei#575]
Hi, I'm very new to the code and trying to get the best parallelization with the machine I'm working on.

I'm simulating particle driven PWFA and the machine specs are the following:

CPU(s):                      8
Thread(s) per core:    2
Core(s) per socket:    4
Socket(s):                  1

I tried different combinations of MPI processes and  OpenMP threads to evaluate performances, fixing the product of processes and threads to get 8 total threads every time (1MPI and 8OMP, 2MPI and 4OMP and so on).

However, before the first PIC loop I always get this warning:

"The following `push time` assumes a global number of XXX cores (hyperthreading is unknown)"

with XXX sometimes different from the expected 8 total threads. Am I missing something?
You can find some examples of the described behavior below, from the fastest (A) to the slowest (C) configuration in terms of PIC execution time:

(A)    4MPI , 2OMP  ->  XXX = 8  
(B)    2MPI , 4OMP  ->  XXX = 4  
(C)    1MPI , 8OMP  ->  XXX = 2  

Thank you for the attention,

Andrea

**Comment by mccoys** [comment]
8 threads is not likely to be optimal. If you have 4 cores, it is generally better to have 4 threads.

The warning is always present. It should be equal to nMPI x nOMP. Check in the output (before the warning) where it indicates the number of MPIs and threads.

**Comment by afrawake** [comment]
Thank you for the quick answer, I will follow your suggestion. 
Just making a little practice with parallelization before having access to a bigger machine and wanted to check if I was messing up with something.

The output before the warning shows the right number of MPI processes and OMPthreads per MPI, but the number shown in the warning itself doesn't always equal nMPI x nOMP, as I reported in the first message.
The reason I tried more MPI processes was to speed up relativistic poisson init.

**Comment by mccoys** [comment]
Could you explain how you set the number of mpis and threads? Could you paste here an example of the log?

**Comment by Z10Frank** [comment]
@mccoys indeed, as @afrawake said, increasing the number of MPI processes makes the initial relativistic Poisson solver quicker, since it is parallelized only with MPI and not OpenMP at the moment. Parallelizing it also with OpenMP should become a feature request.

**Comment by afrawake** [comment]
OpenMP threads number is set with export OMP_NUM_THREADS=...
MPI processes number is set with mpirun -n ... before ./smilei my_namelist.py
The initial log looks coherent with the values I set. If I type

export OMP_SCHEDULE=dynamic
OMP_NUM_THREADS=4
mpirun -n 2 ./smilei my_namelist.py

the log reports

MPI_THREAD_MULTIPLE enabled
Number of MPI processes: 2
Number of threads per MPI process : 4
 
Number of patches: 64 x 2
Number of cells in one patch: 11 x 160
Dynamic load balancing: never

and that looks fine (I checked for all other combinations of MPI, OMP).
Instead the warning message reports

 "[WARNING](0) src/Params/Params.cpp:1414 (print_timestep_headers) The following push time assumes a global number of 4 cores (hyperthreading is unknown)"

More examples are the (A) and (C) cases in my first message.

Made several trials and the best configuration for my machine specs looks the one with 2 patches along r axis (simulating in AM), 4 MPI and 2 OpenMP threads.

Thank you again,

Andrea
